torch/fx/experimental/symbolic_shapes.py

```diff
@@ -5344,22 +5344,37 @@ def _maybe_evaluate_static(
 
         # axioms with compute hint NYE
         assert not compute_hint or not axioms
-
         expr = self.simplify(expr)
 
         if compute_hint:
             expr = expr.xreplace(self.var_to_val).xreplace(self.unbacked_var_to_val)
 
         expr = canonicalize_bool_expr(expr)
 
+        def try_simplify_floor_div(axioms) -> None:
+            new_items = {}
+            for k, v in axioms.items():
+                # A FloorDiv in implications could have became CleanDiv at this point, due to new facts
+                # to the shapeEnv. This handles such issue but its not ideal. This is the only expression
+                # simplification that depends on the global state of shape env.
+                # TODO try to get rid of CleanDiv since it breaks the invariant thats simplifications of sympy
+                # expressions only depend on the expression itself.
+                if k.has(FloorDiv):
+                    new_items.update({self.simplify(k): v})
+            axioms.update(new_items)
```

Comment:
I think a slight optimization is that you only have to do these updates when a new divisibility constraint is added; e.g., if I keep doing CleanDiv over and over again, I don't have to keep trying to simplify here (I think it's relatively simple to make a synthetic test that would keep hitting `try_simplify_floor_div` here).

Reply:
good idea let me do that

Reply:
here is a situation were that can happen and be expensive, 
```
        from sympy import Eq

        test1 = Eq(FloorDiv(a.node.expr, 3), 1)
        test2 = Eq(CleanDiv(a.node.expr, 3), 1)

        self.assertTrue(shape_env.evaluate_expr(test1))
        self.assertEqual(shape_env._maybe_evaluate_static(test2), None)

        # After this FloorDiv(a, 3) is simplified to CleanDiv(a, 3)
        shape_env.defer_runtime_assert(Eq(Mod(a, 3), 0), " test")
        self.assertEqual(test2, shape_env.simplify(test1))
        #25692715671
        for i in range(1, 1000):
            # add unrelated change to shape env.
            a = shape_env.create_unbacked_symint()
            shape_env.defer_runtime_assert((a // 3 == 1).node.expr, " test")

        from torch._C import _instruction_counter
    
        start = _instruction_counter.start()

        print(start)
        for i in range(1, 1000):
            # add unrelated change to shape env.
            a = shape_env.create_unbacked_symint()
            shape_env.defer_runtime_assert((a== 1).node.expr, " test")
            a = shape_env.create_unbacked_symint()
            shape_env._maybe_evaluate_static(CleanDiv(a.node.expr, i))

        end = _instruction_counter.end(start)
        print(end)
```